### ISSUE TITLE
Feature Request: Provide a digit limit option of random seed value

### ISSUE THREAD

**Issue by BigAlanM** [toolbox4minecraft/amidst#27]
Currently the MC pseudo random number generator produces a 64 bit random number which is represented as a signed quad integer (-9,223,372,036,854,775,808 thru 9,223,372,036,854,775,807).
Those are 19 digit numbers. 
There are almost 9 times as many 19 digit numbers as there are 18 digit number and there are 10 times as many 18 digit numbers as there are 17.
With the current PRNG the odds are such that 99.99% of the seeds produced will be 16 or more digits long.
The Java routine hashcode() hashes text strings entered as seeds to 32 bit signed integers
(-2,147,483,648 thru 2,147,483,647).
This leaves a large group of seeds in between these two ranges that are never seen by players unless they manually enter them manually.
I tried this for a while. It's a PITA.

I'd really like to see the Ctrl-R function have an option that limits the maximum number of digits of the random seed.

There's a refinement to this that is REALLY esoteric.  It seems that the world generator for underground features such as caves, ravines, and IIRC ore placement ONLY uses the lower 48 bits of the 64 bit seed.  That leaves the upper 16 bits to cause changes in biomes, and surface features.
This means that for every configuration of caves and ravines there are just over 65,000 different surface worlds.
What would be a nice feature would be to "Lock" the lower 48 bits and then produce random maps of worlds that only differ in the upper 16 bits of the seed.

**Comment by stefandollase** [comment]
I see that the second request can have a use case: You can first search a seed with the correct structures. Afterwards, you can search for the correct set of biomes. So, this might be implemented. For example SHIFT+CTRL+R could generate a new world where only the upper 16 bits of the seed are changed randomly.

However, I do not really understand your first point. This is just how numbers work. I do not see the implications of this for the generated world, or is the issue that the long random number is hard to type in correctly? In this case, I can suggest copy and paste :-P In general, all 64-bit numeric seeds are chosen evenly random by Amidst. Maybe, you can elaborate your description?

**Comment by BigAlanM** [comment]
Yes, the PRNG produces 64vbit signed integers spread evenly across the range.
Let's just look at 1 and 2 digit numbers. I'll only mention positive numbers the same is true for negatives, the proportions stay the same.
There are 9 single digit numbers, 1 through 9.  (Zero is a special case and doesn't affect the ratios)
There are 90 two digit numbers, 10 through 99.
Therefore if you have a perfect PRNG limited to 1 through 99 10 times more of them will be two digit than single digit.
If you add one more digit there will be 900 possible 3 digit results and now the chances of getting a random 1 digit result is down to 1%.

Back to what I stated in the first comment with a PRNG that has upper limit of 19 digit numbers, even though the results are smoothly distributed over the range, 99.99% of the generator's outputs will have greater than 15 digits.
That is, they will be in the range of  +/- 9,223,372,036,854,775,807 down to +/-1,000,000,000,000,000.

Except for the numbers in the range of -2,147,483,648 thru 2,147,483,647 that hashcode() produces when a text string is entered. (and not all of those are possible to produce with hashcode()) the chances of a seed being in the range of 999,999,999,999,999 down to -999,999,999,999,999 are extremely small.  10 times smaller for each digit you get closer to the upper values of hashcode() output.

BTW 999,999,999,999,999  is a 50 bit binary number.
In Hex it's: 
3 8D7E A4C6 7FFF
in binary it's:
11 1000 1101 0111 1110 1010 0100 1100 0110 0111 1111 1111 1111
I have NO idea why I brought that up.

**Comment by stefandollase** [comment]
I still do not get the point. The problem is not the math. I don't understand what you want to have changed and why.

You described it pretty well: Most of the numbers have many digits. This is, because when you add a digit, the amount of numbers that can be represented is multiplied by 9. However, I think your conclusion is wrong. It is not the case, that a given number with few digits is chosen less often than a given number with many digits. What is true is, that it is less likely to choose a number with few digits than it is to choose a number with many digits. The only reason for this is, that there are much more numbers with many digits. That is what I meant when I said:
> This is just how numbers work.
Since this the normal behavior, when a number is chosen evenly random, I do not see any reason to change this. You also did not come up with a reason for this until now, that makes sense to me. Is the reason the copy&paste thing? If yes, can you describe what exactly you mean by it?

**Comment by maruohon** [comment]
I have to admit, I also didn't understand what the actual goal of this limit would be.  Other than the 48-bit limit, which I didn't know about prior to this, that would be a really cool feature.

Let's say there is a 10-digit limited mode, and the random seed it chooses is for example 8 digits long. How exactly is that "easier" or less of a PITA than a 12-digit seed? I'm assuming the PITA part is just transferring the seed of an interesting looking world (in regards to the biomes) from Amidst to Minecraft itself? Wouldn't you copy paste the 8-digit seed as well?

And how are seeds between the parsed/hashed text strings and the "most likely range for a pure random seed" never seen by players? If they generate a random world with no pre-defined seed, the resulting random seed can be in that range, although it to be in that range is less likely than it being in the upper range due to the above mentioned reasons. But why does it matter? There are so many values in any of the ranges mentioned here that why would this make a difference in any way, where the seed value lies? They are all equally likely to be awesome or boring. If you want a specific known world, then you need to use that specific seed anyway.

**Comment by BigAlanM** [comment]
The PITA is in getting a group of 10 digit (for example) seeds displayed in Amidst.  Using another program to generate a random list or by just using Ctrl-C on a randomly produce map then using Ctrl-H, Ctrl-V, and then deleting all but 10 digits is quite time consuming when someone wishes to look at a large number of 10 digit seed worlds.

**Comment by maruohon** [comment]
Well my point was, WHY do you want to look at specifically seeds that have less digits? I'm assuming you are searching for nice worlds based on random seeds, so why do you want to limit the seed "type"?
